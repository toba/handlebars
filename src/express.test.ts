import '@toba/test';
import * as path from 'path';
import Express = require('express');
import request = require('supertest');
import { ExpressHandlebars } from '.';
import { HttpStatus } from '@toba/node-tools';

const viewPath = path.join(__dirname, '__mocks__', 'views');
const app = Express();
const items = [{ name: 'one' }, { name: 'two' }, { name: 'three' }];

beforeAll(() => {
   const ehb = new ExpressHandlebars();
   app.set('views', viewPath);
   app.set('view engine', ehb.fileExtension);
   app.engine(ehb.fileExtension, ehb.renderer);
});

function makeRoute(path: string, viewName: string, layout?: string) {
   app.get(path, (_req: Express.Request, res: Express.Response) => {
      res.render(
         viewName,
         { items, title: 'Mockery', layout },
         (err: Error, html: string) => {
            expect(err).toBeNull();
            res.write(html);
            res.end();
         }
      );
   });
}

test('renders body with partial within layout', async () => {
   makeRoute('/body-with-partial', 'partial');
   const res = await request(app).get('/body-with-partial');
   expect(res.status).toBe(HttpStatus.OK);
   expect(res.text).toMatchSnapshot();
});

test('renders without layout', async () => {
   makeRoute('/', 'home');
   const res = await request(app).get('/');
   expect(res.status).toBe(HttpStatus.OK);
   expect(res.text).toMatchSnapshot();
});

test('renders within default layout', async () => {
   makeRoute('/home', 'home');
   const res = await request(app).get('/home');
   expect(res.status).toBe(HttpStatus.OK);
   expect(res.text).toMatchSnapshot();
});

test('renders within specified layout', async () => {
   makeRoute('/other', 'home', 'other');
   const res = await request(app).get('/other');
   expect(res.status).toBe(HttpStatus.OK);
   expect(res.text).toMatchSnapshot();
});

test('renders data', async () => {
   makeRoute('/bare-data', 'data');
   const res = await request(app).get('/bare-data');
   expect(res.status).toBe(HttpStatus.OK);
   expect(res.text).toMatchSnapshot();
});

test('renders data within layout', async () => {
   makeRoute('/data', 'data');
   const res = await request(app).get('/data');
   expect(res.status).toBe(HttpStatus.OK);
   expect(res.text).toMatchSnapshot();
});
