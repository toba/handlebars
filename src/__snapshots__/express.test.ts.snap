// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders body with partial within layout 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" /> 
</head>

<body>
   <nav>
   Navigation
</nav> <section>
   <h4>Don't eat bugs</h4></section> <footer>
   Footer
</footer>
</body>

</html>"
`;

exports[`renders data 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" /> 
</head>

<body>
   <nav>
   Navigation
</nav> <h1>Mockery</h1>
<ol>
   <li>one</li>
   <li>two</li>
   <li>three</li>
</ol> <footer>
   Footer
</footer>
</body>

</html>"
`;

exports[`renders data within layout 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" /> 
</head>

<body>
   <nav>
   Navigation
</nav> <h1>Mockery</h1>
<ol>
   <li>one</li>
   <li>two</li>
   <li>three</li>
</ol> <footer>
   Footer
</footer>
</body>

</html>"
`;

exports[`renders within default layout 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" />    <meta name=\\"content-for\\" />

</head>

<body>
   <nav>
   Navigation
</nav> 
<article>
   <h1>Home</h1>
</article> <footer>
   Footer
</footer>
</body>

</html>"
`;

exports[`renders within specified layout 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" />    <meta name=\\"content-for\\" />

</head>

<body>
   <nav>
   Navigation
</nav> 
<article>
   <h1>Home</h1>
</article> <footer>
   Footer
</footer> Other layout
</body>

</html>"
`;

exports[`renders without layout 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
   <meta value=\\"something\\" />    <meta name=\\"content-for\\" />

</head>

<body>
   <nav>
   Navigation
</nav> 
<article>
   <h1>Home</h1>
</article> <footer>
   Footer
</footer>
</body>

</html>"
`;
